src/app/clusters/groups-server/groups-server.cpp

```diff
@@ -232,141 +261,150 @@ bool emberAfGroupsClusterViewGroupCallback(app::CommandHandler * commandObj, con
+                                      TLV::kTLVType_Array, type));
             {
-                EmberBindingTableEntry entry;
-                status = emberGetBinding(j, &entry);
-                if ((status == EMBER_SUCCESS) && (entry.type == EMBER_MULTICAST_BINDING))
+                GroupDataProvider::GroupMapping mapping;
+                size_t requestCount = 0;
+                size_t matchCount   = 0;
+                ReturnErrorOnFailure(mCommandData.groupList.ComputeSize(&requestCount));
+
+                emberAfGroupsClusterPrint("RX: GetGroupMembership [");
+                if (0 == requestCount)
                 {
-                    if (entry.local == emberAfCurrentEndpoint() && entry.groupId == groupId)
+                    // 1.3.6.3.1. If the GroupList field is empty, the entity SHALL respond with all group identifiers of which the
+                    // entity is a member.
+                    while (mIterator->Next(mapping))
                     {
-                        list[listLen]     = EMBER_LOW_BYTE(groupId);
-                        list[listLen + 1] = EMBER_HIGH_BYTE(groupId);
-                        listLen           = static_cast<uint8_t>(listLen + 2);
-                        count++;
+                        ReturnErrorOnFailure(chip::app::DataModel::Encode(writer, TLV::AnonymousTag, mapping.group));
+                        matchCount++;
+                        emberAfGroupsClusterPrint(" 0x%02" PRIx16, mapping.group);
                     }
                 }
+                else
+                {
+                    while (mIterator->Next(mapping))
+                    {
+                        auto iter = mCommandData.groupList.begin();
+                        while (iter.Next())
+                        {
+                            if (mapping.group == iter.GetValue())
+                            {
+                                ReturnErrorOnFailure(chip::app::DataModel::Encode(writer, TLV::AnonymousTag, mapping.group));
+                                matchCount++;
+                                emberAfGroupsClusterPrint(" 0x%02" PRIx16, mapping.group);
+                                break;
+                            }
+                        }
+                        ReturnErrorOnFailure(iter.GetStatus());
+                    }
+                }
+                emberAfGroupsClusterPrintln("]");
+
+                // 1.3.6.3.1: If the GroupList field contains one or more group-id’s but does not contain any group
+                // of which the entity is a member, the entity SHALL only respond if the command is unicast
+                VerifyOrReturnError((0 == requestCount) || (matchCount > 0) ||
+                                        (emberAfCurrentCommand()->type == EMBER_INCOMING_UNICAST),
+                                    CHIP_ERROR_INVALID_DATA_LIST);
```

Comment:
This code block still does not make sense...

Reply:
I actually agree, but this is my best interpretation of the spec. Unfortunately, the spec is ambiguous at this point, and says what should happen when the condition is true, but not what should happen when the condition is false, so I aborted the response with an error.

Reply:
My point is that in Matter this is handled on the IM layer, not in cluster impls.  This block should just not exist.